internal/locate/region_cache_test.go

```diff
@@ -2519,3 +2520,88 @@ func BenchmarkInsertRegionToCache2(b *testing.B) {
 		cache.insertRegionToCache(region, true, true)
 	}
 }
+
+func (s *testRegionCacheSuite) TestBatchScanRegions() {
+	// Split at "a", "b", "c", "d", "e", "f", "g"
+	// nil --- 'a' --- 'b' --- 'c' --- 'd' --- 'e' --- 'f' --- 'g' --- nil
+	// <-  0  -> <- 1 -> <- 2 -> <- 3 -> <- 4 -> <- 5 -> <- 6 -> <-  7  ->
+	regions := s.cluster.AllocIDs(7)
+	regions = append([]uint64{s.region1}, regions...)
+
+	peers := [][]uint64{{s.peer1, s.peer2}}
+	for i := 0; i < 7; i++ {
+		peers = append(peers, s.cluster.AllocIDs(2))
+	}
+
+	for i := 0; i < 7; i++ {
+		s.cluster.Split(regions[i], regions[i+1], []byte{'a' + byte(i)}, peers[i+1], peers[i+1][0])
+	}
+
+	check := func(scanRanges, cachedRanges []kv.KeyRange, afterCacheLoad func(), expected []uint64) {
+		s.cache.clear()
+		// fill in the cache
+		for _, r := range cachedRanges {
+			for {
+				loc, err := s.cache.LocateKey(s.bo, r.StartKey)
+				s.Nil(err)
+				if loc.Contains(r.EndKey) || bytes.Equal(r.EndKey, loc.EndKey) {
+					break
+				}
+				r.StartKey = loc.EndKey
+			}
+		}
+		if afterCacheLoad != nil {
+			afterCacheLoad()
+		}
+		scannedRegions, err := s.cache.BatchLocateKeyRanges(s.bo, scanRanges)
+		s.Nil(err)
+		s.Equal(len(expected), len(scannedRegions))
+		actual := make([]uint64, 0, len(scannedRegions))
+		for _, r := range scannedRegions {
+			actual = append(actual, r.Region.GetID())
+		}
+		s.Equal(expected, actual)
+	}
+
+	toRanges := func(keys ...string) []kv.KeyRange {
+		ranges := make([]kv.KeyRange, 0, len(keys)/2)
+		for i := 0; i < len(keys); i += 2 {
+			ranges = append(ranges, kv.KeyRange{StartKey: []byte(keys[i]), EndKey: []byte(keys[i+1])})
+		}
+		return ranges
+	}
+
+	// nil --- 'a' --- 'b' --- 'c' --- 'd' --- 'e' --- 'f' --- 'g' --- nil
```

Comment:
We could add test to assume the region information merge happens at
The beginning
2. The middle
3. The end

Or there are stale region information in the region cache from these positions.

Reply:
The cases are difficult to test in the integration `BatchLocateKeyRanges`, I add the coverage in `TestBatchScanRegionsMerger` and `TestSplitKeyRanges`.